### DISCUSSION TITLE
BinaryPrimitives endianness-specific methods contain confusing branches

### DISCUSSION THREAD

**Discussion by siegfriedpammer** [dotnet/roslyn#74892] [Q&A] [answered]
See the original bug report in the ILSpy repository: https://github.com/icsharpcode/ILSpy/issues/3166

It looks like some branch optimization was only done halfway or is there any specific reason why the method call in the else branch was removed?
- Is `MethodImplOptions.AggressiveInlining` honored by Roslyn? If that's the case, then
- why isn't the condition check and branch eliminated as well?

The IL in question contains a branch instruction with a relative target-offset of 0, which basically means "don't jump at all", so ILSpy is decompiling/displaying the code correctly.

`BitConverter.IsLittleEndian` is marked with the `IntrinsicAttribute` but that doesn't seem to change Roslyn's behavior... I even checked with SharpLab and applying `MethodImplOptions.AggressiveInlining` in Release mode does not cause any call to be removed.

https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwAYACDHAOgCUBXGBKMAU0oGEB7MAB1WYCcAygIBuUAMbMAzgG5CJclToMmrAJIMBHbsP5jJs+QGZyWUm1IBvQqVukA2gFlmCABYcAJmp4oAFM7dPb24UAHluRg4YKUoAQQBzeP5pKSgRZg0UWFh4gEoAXRs7DBMKADZSWARSamYILwYjLAAhKHiAURgPKAgYX1r60JgUAE9Bbl6AHmARhGYAPlIpDlp+SVyi22sCO13yAHZSVoROGHT+Of5KNSkAGSgEBBRmTu7e0gB+Tb3d2vOpF5dHowGApfzMMAcfgjRwQfhSVwQFA0OoeSZVea+Zarda5Uggb4/WzOSHQ2HwxHIgZojFYlZrZi5OQ7OwAX2+Thc7i8PnBgR5IXCkWicUSySkqXSmWyMDyhRZtm4/DSEDmigqVRqzH+gLeILBmpESNojNIAF5FkaUCbmaygA

The generated code contains 3 method calls, as expected.

Would like to know if this should be reported as a bug or if there is a different explanation for this behavior.

**Comment by huoyaoyuan** [comment, marked as answer]
> * Is `MethodImplOptions.AggressiveInlining` honored by Roslyn?

No, it's for JIT.

> `BitConverter.IsLittleEndian` is marked with the `IntrinsicAttribute` but that doesn't seem to change Roslyn's behavior...

`[Intrinsic]` is for JIT and roslyn know nothing about it.
In fact, roslyn can't even see the `[Intrinsic]` or `[MethodImplOptions.AggressiveInlining]` attributes since they are only present in implementation, not the reference assembly.

IL is expected to contain calls and branches around intrinsic functions. They will be constant folded by JIT.

You should view JIT asm in sharplab. There's only one branch for range checking in this case.

**Reply by siegfriedpammer** [reply]
Ah, so it's ILLink. Thanks.
